content/posts/csv-boilerplate-for-go.md

```diff
@@ -0,0 +1,307 @@
+
+```go
+type Car struct {
+	Car          string
+	MPG          float64
+	Cylinders    int64
+	Displacement float64
+	Horsepower   float64
+	Weight       float64
+	Acceleration float64
+	Model        int64
+	Origin       string
+}
+```
+
+My first naive attempt is usually to write some code that defines the headers and struct mapping in a way that is separate from the CSV encoding/decoding itself. An example of that can be seen below:
+
+```go
+var carHeaders = []string{
+  "Car",
+  "MPG",
+  "Cylinders",
+  // remaining columns ...
+}
+
+func (c *Car) UnmarshalRecord(record []string) (err error) {
+	if got, want := len(record), len(carHeaders); got != want {
+		return fmt.Errorf("bad column number: got=%d want=%d", got, want)
+	}
+	c.Car = record[0]
+	if c.MPG, err = strconv.ParseFloat(record[1], 64); err != nil {
+		return fmt.Errorf("column=%q: %w", carHeaders[1], err)
+	} else if c.Cylinders, err = strconv.ParseInt(record[2], 10, 64); err != nil {
+		return fmt.Errorf("column=%q: %w", carHeaders[2], err)
+	}
+	// remaining columns ...
+	return nil
+}
+
+func (c *Car) MarshalRecord() ([]string, error) {
+	return []string{
+		c.Car,
+		fmt.Sprintf("%f", c.MPG),
+		fmt.Sprintf("%d", c.Cylinders),
+		// remaining columns ...
+	}, nil
+}
+```
+
+This works, but adding, removing, or reordering any of the columns requires us to modify our code in 4 to 5 different places, which we wanted to avoid. So let's try some abstraction that allows us to declaratively specify our columns and how to marshal/unmarshal them:
+
+```go
+type carColumn struct {
+	Name           string
+	Type           string
+	UnmarshalValue func(*Car, string) error
+	MarshalValue   func(*Car) (string, error)
+}
+
+var carColumns = []carColumn{
+	{
+		"Car",
+		"STRING",
+		func(c *Car, val string) error {
+			c.Car = val
+			return nil
+		},
+		func(c *Car) (string, error) {
+			return c.Car, nil
+		},
+	},
+	{
+		"MPG",
+		"DOUBLE",
+		func(c *Car, val string) (err error) {
+			c.MPG, err = strconv.ParseFloat(val, 64)
+			return
+		},
+		func(c *Car) (string, error) {
+			return fmt.Sprintf("%f", c.MPG), nil
+		},
+	},
+	{
+		"Cylinders",
+		"INT",
+		func(c *Car, val string) (err error) {
+			c.Cylinders, err = strconv.ParseInt(val, 10, 64)
+			return
+		},
+		func(c *Car) (string, error) {
+			return fmt.Sprintf("%d", c.Cylinders), nil
+		},
+	},
+	// remaining columns ...
+}
+```
+
+This is of course a bit verbose, but on the other hand it solves our problem. Reordering our columns is now a single cut & paste operation, and adding or removing a column has gotten a lot easier as well. The approach even gives us a convenient way to put the `Type` information found in the second line of the CSV file, so we can easily duplicate this quirk when writing our own CSV files later on.
+
+Of course we still need to update our `UnmarshalRecord` and `MarshalRecord` from earlier. The good news is that we're unlikely to ever have to modify this code again:
```

Comment:
```suggestion
We still need to update our `UnmarshalRecord` and `MarshalRecord` from earlier. The good news is that we're unlikely to ever have to modify this code again:
```

Reply:
This one slipped through the cracks: 7afdc00a63ef5cb8036704db1f7d03bda9732b1f